cpp/src/arrow/c/bridge.cc

```diff
@@ -663,6 +672,112 @@ Status ExportRecordBatch(const RecordBatch& batch, struct ArrowArray* out,
   return Status::OK();
 }
 
+//////////////////////////////////////////////////////////////////////////
+// C device arrays
+
+Status ValidateDeviceInfo(const ArrayData& data, DeviceType* device_type,
+                          int64_t* device_id) {
```

Comment:
Not a big deal ATM, but a `DeviceInfo` struct might be good have (not just here, but in general), e.g:
```c++
struct DeviceInfo {
  explicit DeviceInfo(DeviceType type = DeviceType::UNKNOWN, int64_t id = -1)
      : type(type), id(id) {}
  DeviceType type;
  int64_t id;
};
```

Reply:
Or at least return `Result<std::pair<DeviceType, int64_t>>` here...

Reply:
That brings me back to my original implementation vs this tail recursion version that @felipecrv suggested haha. 

This is the internal recursive call, below it is the intended function to call which does return `Result<std::pair<DeviceType, int64_t>>`

Reply:
If you don't like passing a `Result` recursively, you can always build a struct with a method in it...

Reply:
the `Result` isn't recursive, the recursion just passes the pointers to the device type and id